Added gravity into jump height calculation

braem opened the pull request:
Closes #539 when merged into `develop`

Found the factors based on what the jump height should be at the gamemode's default gravity & added `sv_gravity` into the calculation.
IE. found `x^2` where `jumpHeight = x^2 / (2 * baseGrav)`. For ahop, `jumpHeight ~ 21.333333` and `baseGrav = 600`.
I just used the squared value since the calculation involves squaring anyhow.

- **ahop** -> jump should be `21.333333..` at base gravity (`600`). This gives `25600` (not squared is `160`).
- **TF2** -> jump should be ~`52.200625` at base gravity (`800`). This gives `83521` (not squared is `289.0`).
- **Base** -> jump should be `57` at base gravity (`800`). This gives `91200` (not squared is `301.99337741083..`).

### Checklist
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

braem commented:
Was not calculating the square precisely.. will tweak those values.

Gocnak commented:
I believe steam HL2 uses a factor of 160.0f: https://github.com/momentum-mod/game/blob/develop/mp/src/game/shared/gamemovement.cpp#L2432

braem commented:
> 
> 
> I believe steam HL2 uses a factor of 160.0f: https://github.com/momentum-mod/game/blob/develop/mp/src/game/shared/gamemovement.cpp#L2432

Sorry about that, caused myself some confusion over having the wrong number, I updated the PR comment. The numbers in this PR are correct. 
The calculation for ahop would be `160 * 160 / (2 * grav)`, but I used the squared value instead (`25600`).

Decided to use the squared values as that's what we get when determining the factor from the jump height (`x^2 = jumpHeight * (2 * baseGrav`).

chrb22 commented:
Wouldn't it be easier to instead have a GetJumpVelocity() and then remove the velocity calculation in CheckJumpButton() instead of this?

Gocnak commented:
Yeah does changing gravity affects this? The `2.0f * sv_gravity.GetFloat()` from `CheckJumpButton` would effectively cancel out the `/ (2.0f * sv_gravity.GetFloat())` in the `GetJumpHeight()`?

Effectively making it just return the number, which is then square rooted every tick to be the jump factors...

braem commented:
> 
> 
> Yeah does changing gravity affects this? The `2.0f * sv_gravity.GetFloat()` from `CheckJumpButton` would effectively cancel out the `/ (2.0f * sv_gravity.GetFloat())` in the `GetJumpHeight()`?
> 
> Effectively making it just return the number, which is then square rooted every tick to be the jump factors...

It does work on different gravities yes. Removing the `2.0f * sv_gravity.GetFloat()` on both sides has  the same effect. I suppose neither of these are needed as you can just set `fJumpFactor` to `160`/`289`/`301.99337741083..` and it'll work.

braem commented:
I removed those `2.0f * sv_gravity.GetFloat()`'s from the calculation and renamed `GetJumpHeight()` to `GetJumpFactor()`.

Tested and works on all gamemodes.

Gocnak commented:
So now does changing gravity affect jump height...?

chrb22 commented on mp/src/game/shared/momentum/mom_system_gamemode.cpp:
```diff
@@ -50,6 +50,13 @@ void CGameModeBase::SetGameModeVars()
+{
+    // height == 57 on 800 gravity
+    // x = sqrt(height * 2 * gravBase) = sqrt(57 * 2 * 800) ~ 301.99337741083
+    return 301.99337741083f;
```
Would be nicer to just write sqrt(2.0f * 800.0f * 57.0f) instead as the value should be evaluated at compile time.

braem replied:
It would make more sense yeah, this is just an optimization to avoid having to square root & do other calculations.

chrb22 replied:
I think the compiler is smart enough to do the calculation when compiling.

braem replied:
You're right :)

chrb22 commented:
> So now does changing gravity affect jump height...?

It will.

braem commented:
> 
> 
> So now does changing gravity affect jump height...?

Yup! I can record a short vid if you are skeptical.

braem commented:
Fixup'd the commit again with the jump factor calculation instead of just having the value.
Kept one comment just in case someone has to revisit that.

chrb22 commented on mp/src/game/shared/momentum/mom_system_gamemode.cpp:
```diff
@@ -120,6 +126,11 @@ void CGameMode_RJ::SetGameModeVars()
 
+float CGameMode_RJ::GetJumpFactor()
+{
+    return sqrtf(52.200625f * 2.0f * 800.0f);
```
Using the actual jump velocity for jump and ahop would make more sense since they are already defined using a velocity and not a jump height like CS:S.

braem replied:
Ah, changed those back then :)

chrb22 commented:
@Gocnak Didn't CS:S also use = for jumping while ducking? Because might as well remove !g_pGameModeSystem->IsCSBasedMode() while we (brae) is doing jump velocity stuff.

Gocnak commented:
@chrb22 A while ago, mev made changes to make jumping for CS:S have consistent 64s. Having it force the `else` branch is necessary for those changes to still work, and would otherwise mess up that effort.
